src/zenml/integrations/modal/orchestrators/modal_orchestrator.py

```diff
@@ -0,0 +1,389 @@
+                        timeout=settings.timeout or self.config.timeout,
+                        environment_name=settings.modal_environment
+                        or self.config.modal_environment,
+                        synchronous=settings.synchronous
+                        if hasattr(settings, "synchronous")
+                        else self.config.synchronous,
+                    )
+                )
+        except Exception as e:
+            logger.error(f"Pipeline execution failed: {e}")
+            logger.info("Check Modal dashboard for detailed logs")
+            raise
+
+        logger.info("Pipeline execution completed successfully")
+
+        return None
+
+    async def _execute_pipeline_sandbox(
+        self,
+        app_name: str,
+        zenml_image: Any,
+        entrypoint_command: List[str],
+        deployment: "PipelineDeploymentResponse",
+        execution_mode: str,
+        run_id: Optional[str] = None,
+        gpu_values: Optional[str] = None,
+        cpu_count: Optional[int] = None,
+        memory_mb: Optional[int] = None,
+        cloud: Optional[str] = None,
+        region: Optional[str] = None,
+        timeout: int = 86400,
+        environment_name: Optional[str] = None,
+        synchronous: bool = True,
+    ) -> None:
+        """Execute pipeline using Modal sandbox.
+
+        Args:
+            app_name: Name of the Modal app
+            zenml_image: Pre-built ZenML Docker image for Modal
+            entrypoint_command: Command to execute in the sandbox
+            deployment: Pipeline deployment for tagging
+            execution_mode: Execution mode for tagging
+            run_id: Pipeline run ID for tagging
+            gpu_values: GPU configuration string
+            cpu_count: Number of CPU cores
+            memory_mb: Memory allocation in MB
+            cloud: Cloud provider to use
+            region: Region to deploy in
+            timeout: Maximum execution timeout
+            environment_name: Modal environment name
+            synchronous: Whether to wait for completion
+        """
+        # Create persistent app (will reuse if exists)
+        app = modal.App.lookup(
+            app_name, create_if_missing=True, environment_name=environment_name
+        )
+
+        logger.info(f"Using Modal app: {app_name}")
+
+        logger.info("Creating sandbox for pipeline execution")
+
+        # Generate tags for the sandbox
+        sandbox_tags = generate_sandbox_tags(
+            pipeline_name=deployment.pipeline_configuration.name,
+            deployment_id=str(deployment.id),
+            execution_mode=execution_mode,
+            run_id=run_id,
+        )
+        logger.info(f"Sandbox tags: {sandbox_tags}")
+
+        with modal.enable_output():
+            # Create sandbox with the entrypoint command
+            # Note: Modal sandboxes inherit environment from the image
+            sb = await modal.Sandbox.create.aio(
```

Comment:
why is async interface used to create sandboxes? instead of just `modal.Sandbox.create`

Reply:
- await modal.Sandbox.create.aio() properly integrates with the asyncio event loop
    - modal.Sandbox.create() would make blocking system calls that don't yield control back to the event

  In this specific case, since we're using asyncio.run() and only creating one sandbox at a time in
  pipeline mode, the practical difference might seem minimal. However:

  - Per-step mode: When creating multiple sandboxes concurrently, async is crucial
  - Future extensibility: Allows for potential concurrent operations
  - Modal's design: Modal's async API is optimized for better performance and proper resource management